### ISSUE TITLE
Consider requiring PKCE

### ISSUE THREAD

**Issue by michielbdejong** [remotestorage/spec#176]
There seems to be some progress in general opinion about implicit grant flow best practices, where probably we should require https://www.oauth.com/oauth2-servers/pkce/ in how the remoteStorage spec uses OAuth Implicit Grant.

https://tools.ietf.org/id/draft-parecki-oauth-browser-based-apps-02.txt
https://medium.com/oauth-2/why-you-should-stop-using-the-oauth-implicit-grant-2436ced1c926
https://www.google.com/search?q=implicit+flow+problems

**Comment by skddc** [comment]
Just completing the links: the current draft of the BCP can be found at https://datatracker.ietf.org/doc/draft-ietf-oauth-browser-based-apps/ (moved to a new name)

**Comment by michielbdejong** [comment]
cc @fkooman

**Comment by fkooman** [comment]
Yeah, it would be best to switch to authorization code profile and use PKCE. That's what I've been doing for other projects, i.e. support [RFC8252](https://tools.ietf.org/html/rfc8252) "OAuth 2.0 for Native Apps". This draft @skddc refers to is very similar.

Specifically relevant for RS: https://tools.ietf.org/html/draft-ietf-oauth-browser-based-apps-03#section-6.3

**Comment by skddc** [comment]
This draft mentions requirements for keeping implicit grant flow (but generally recommends not using it anymore): https://tools.ietf.org/html/draft-ietf-oauth-security-topics-15

**Comment by elee1766** [comment]
@raucao hi! first of all, thank you for all the work on remotestorage. this is a really cool project and i see so much value in it.

i was curious what would you say the time line on the migration to version 2.0 is?

the reason i am asking is because i would really like to finalize the golang remote storage client and server library i am working on (along with a proof of concept implementation which uses the library) https://gitlab.com/tuxpaint/remotestorage

i encountered an issue when i actually tried using it, as most modern OIDC providers actually do not support implicit grant at all, and require PKCE.

I'm wondering also what your advice for me might be - should i just try to implement the up-in-the-air remoteStorage 2.0 with PKCE in this library for now?

**Comment by raucao** [comment]
Hey @elee1766, great to hear about your work on a Go implementation, and even for both client and server!

If you want your implementation to be used by anything in the real world and in the foreseeable future, it will have to be RS 1.0, since 2.0 is only a long-term future plan, and we need to get everyone migrated to 1.0 with PKCE first.

Personally, I would recommend to start with the current spec and then add PKCE as soon as the reference client supports it (or has a working PR for it at least).

What is the relevance of OIDC providers to your implementation, if I may ask?

**Comment by elee1766** [comment]
@raucao i want to be able to delegate authentication of the remotestorage to a self hosted oidc provider (zitadel, pocketid) so that i can create a server implementation that does not involve user authentication and management, as that is a separate problem. 

so far it seems the easiest way to create a proper remotestorage 1.0 server is to have to do the user management and authentication myself? 

maybe im doing something all wrong though.

**Comment by raucao** [comment]
Serving Webfinger records is required for the user address *domain*, but the RS server implementation can get away with delegating this, too.

Delegating auth entirely requires that you can still verify arbitrary access scopes, and yes, the provider would have to support Implicit Grant flow for the foreseeable future. You could try to split up the responsibilities and only do authentication, i.e. login, via the OpenID Connect provider, but then do the actual RS OAuth for the account in your implementation.

This way, you wouldn't have to worry about any OpenID Connect providers having to be directly compatible with remoteStorage. But you would obviously need some kind of UI for managing connected apps for the user. However, some kind of app dashboard showing connected apps is usually a nice thing to have anyway.
